// testarmada-magellan-local-executor/lib/executor.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _child_process = require("child_process");

exports.default = {
  /*eslint-disable no-unused-vars*/
  setupRunner: function setupRunner() {
    var mocks = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;

    return new Promise(function (resolve) {
      resolve();
    });
  },

  /*eslint-disable no-unused-vars*/
  teardownRunner: function teardownRunner() {
    var mocks = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;

    return new Promise(function (resolve) {
      resolve();
    });
  },

  setupTest: function setupTest(callback) {
    callback();
  },

  teardownTest: function teardownTest(info, callback) {
    callback();
  },

  execute: function execute(testRun, options) {
    var mocks = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : null;

    var ifork = _child_process.fork;

    if (mocks && mocks.fork) {
      ifork = mocks.fork;
    }

    return ifork(testRun.getCommand(), testRun.getArguments(), options);
  },

  summerizeTest: function summerizeTest(magellanBuildId, testResult, callback) {
    callback();
  }
};
